libs/ml/include/ml/core/node.hpp

```diff
@@ -141,6 +143,105 @@ class Node
     return static_cast<bool>(cached_output_status_ == CachedOutputState::VALID_CACHE);
   }
 
+  virtual uint64_t ForwardPassChargeCost()
+  {
+    static constexpr uint64_t my_default_cost = 100;  // TODO(VH): Calculate/measure me
+    // TODO(VH): call OpForwardCost() of the Op - we need to know Data shape to do this.
+    if (input_nodes_.empty())
+    {
+      FETCH_LOG_INFO("Node", " Input node reached: default cost " +
+                                 std::to_string(my_default_cost) + " added.");
+      return my_default_cost;
```

Comment:
I'm not 100% sure about this logic. Seems a bit like we're indirectly saying 'is this node holding a placeholder op?'
It's probably fine, but if we ever have more than one op that doesn't need any input nodes we're assuming they should have the same cost. Should they?

Reply:
The thing here is: a placeholder is shapeless, and its shape can be deduced either from input data (which requires passing input data to all estimation chain), or from the shape of next Dense layer. For now I only traverse the Graph backwards, from output node to the input one; deducing shape of a placeholder Input requires one or more step(s) in opposite direction to find nearest Dense layer and use its shape as a reference. I am going to implement this as well but a bit later.

Reply:
Also there is a problem now with Weight Ops that are not Placeholders but do not have any Input nodes, so I guess here we have to check if the node Ops type is Placeholder or not (and if not, no additional charge would be added)

Reply:
Added operation type check.